File: Programmers/python/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    bridge = deque([0] * bridge_length)
    answer = 0
    cur_weight = 0
    truck_idx = 0
    
    # 아직 건널 트럭이 남아있을 때
    while truck_idx < len(truck_weights):
        answer += 1
        
        # 다리에 트럭이 꽉 차있다면 pop
        if len(bridge) >= bridge_length:
            cur_weight -= bridge.popleft()
        
        # 트럭이 다리 위에 올라올 수 있을 때
        if cur_weight + truck_weights[truck_idx] <= weight:
            truck = truck_weights[truck_idx]
            cur_weight += truck
            truck_idx += 1
            bridge.append(truck)
        else:
            bridge.append(0)

    # 마지막 트럭이 진입하자마자 종료되므로 마지막 트럭이 지나갈 때까지의 시간을 추가
    answer += bridge_length

    return answer

File: Programmers/java/다리를 지나는 트럭.java
import java.util.*;

class Solution {
    public int solution(int bridge_length, int weight, int[] truck_weights) {
        Queue<Integer> queue = initQueue(bridge_length);
        
        int curWeight = truck_weights[0];
        queue.add(curWeight);
        int sec = 1;
        int idx = 1;
        
        while(idx < truck_weights.length) {
            // System.out.println(queue);
            int quit = queue.poll();
            curWeight -= quit;
            int next = truck_weights[idx];
            
            // 다음 트럭이 올라갈 수 있는 상황이라면
            if(curWeight + next <= weight) {
                queue.add(next);
                curWeight += next;
                idx++;
            }
            else {
                queue.add(0);
            }
            sec++;
        }
        
        while(!queue.isEmpty()) {
            // System.out.println(queue);
            queue.poll();
            sec++;
        }
        
        return sec;
    }
    
    private Queue<Integer> initQueue(int bridge_length) {
        Queue<Integer> queue = new LinkedList<>();
        for(int i = 0; i < bridge_length - 1; i++) {
            queue.add(0);
        }
        return queue;
    }
}
